src/Multicall3.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.12;
+
+/// @title Multicall3
+/// @notice Aggregate results from multiple read-only function calls
+/// @dev Multicall & Multicall2 backwards-compatible
+/// @author Michael Elliot <mike@makerdao.com>
+/// @author Joshua Levine <joshua@makerdao.com>
+/// @author Nick Johnson <arachnid@notdot.net>
+/// @author Andreas Bigger <andreas@nascent.xyz>
+
+contract Multicall3 {
+    struct Call {
+        address target;
+        bytes callData;
+    }
+
+    struct Call3 {
+        address target;
+        bytes callData;
+        bool requireSuccess;
+    }
+
+    struct Result {
+        bool success;
+        bytes returnData;
+    }
+
+    function aggregate(Call[] memory calls) public returns (uint256 blockNumber, bytes[] memory returnData) {
+        blockNumber = block.number;
+        uint256 length = calls.length;
+        returnData = new bytes[](length);
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(success, "Multicall aggregate: call failed");
+            returnData[i] = ret;
+            unchecked { i++; }
```

Comment:
This is a design decision but generally I like the following format for unchecking more (you can reuse the function for all of your functions):

```solidity
for (uint256 i; i < length; i = _uncheckedInc(i)) {
    // do something that doesn't change the value of i
}

function _uncheckedInc(uint256 i) private pure returns (uint256) {
    unchecked {
        return i + 1;
    }
}
```
> Note that it's important that the call to `_uncheckedInc` is inlined. This is only possible for Solidity versions starting from `0.8.2`.

Reply:
does this result in a JUMP though making it cost more than inlining it this way? cc gas expert @transmissions11

Reply:
It’ll get inlined, I’m not a fan for stylistic reasons but should be no downside

Reply:
from a quick benchmark the function seemed to be a slightly more expensive (maybe my quick benchmark was flawed), but since both are the same I'm going to leave this as-is